### ISSUE TITLE
System.Security.Cryptography.Csp error on UWP

### ISSUE THREAD

**Issue by JamesNK** [dotnet/corefx#10709]
Hi

A user of my library is getting an error when installing the library on NuGet. My library is netstandard1.3.

He is the error:

```
System.Security.Cryptography.Csp 4.0.0 provides a compile-time reference assembly for System.Security.Cryptography.Csp on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm.
```

https://github.com/JamesNK/Newtonsoft.Json.Schema/issues/43

**Comment by bartonjs** [comment]
Presumably you have direct references to RSACryptoServiceProvider in your code.  You should replace them with just using RSA.Create() and the RSA base class; then that compile-time reference should go away.

Of course, we also should have had a version of the library for UWP which just throws exceptions...

**Comment by JamesNK** [comment]
I don't understand. Why it using `RSACryptoServiceProvider` not allowed?

This is what I am doing:

            RSACryptoServiceProvider rsaCryptoServiceProvider = new RSACryptoServiceProvider();
            rsaCryptoServiceProvider.ImportCspBlob(Convert.FromBase64String(PublicKeyCsp));

            valid = rsaCryptoServiceProvider.VerifyData(data, SHA1.Create(), signature);

How do I perform the same logic with `RSA`? As far as I can tell neither ImportCspBlob and VerifyData are not on `RSA`.

**Comment by bartonjs** [comment]
RSACryptoServiceProvider is the implementation of RSA using Windows CAPI.  The APIs it calls are prohibited to UWP applications, and therefore the .NET Core implementation can't work.

The ImportCspBlob method takes a data format which is particular to CAPI, and that is why it is only on RSACryptoServiceProvider.  To be portable you should use the ImportParameters method which takes an RSAParameters structure.  Since the CAPI blob format is documented, you could convert it (one time or live) to RSAParameters to make your code be more portable.  https://github.com/dotnet/corefx/blob/bddba0dcf2decd22900f1a8a0944408d3b852fd5/src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.cs#L1028-L1070 is how we turn CAPI's blob into RSAParameters for RSACryptoServiceProvider.ExportParameters.

**Comment by JamesNK** [comment]
Thanks. Is there an equivalent of RSACryptoServiceProvider.VerifyData? I'm signing a message on the server and I need to verify it on the client. Is there a portable/netstandard way to do that?

**Comment by bartonjs** [comment]
Yep: RSACryptoServiceProvider.VerifyData(byte[] data, object algorithm, byte[] signature) => RSA.VerifyData(byte[] data, byte[] signature, HashAlgorithmName algorithm, RSASignaturePadding.Pkcs1)

Instead of a magic string or a HashAlgorithm object for the algorithm now it's a strongly typed input; and RSA signatures now support a different kind of padding mode (PSS) so we added an input for that.  To be compatible with existing signatures from RSACSP you need to specify RSASignaturePadding.Pkcs1.

(example: `return rsa.VerifyData(data, signature, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);`)

**Comment by JamesNK** [comment]
Thanks again. I missed `RSA.VerifyData` because intellisense is broken in VS for me for netstandard projects, and when I looked for it in intellisense in a regular .NET I missed it again because it was added after 4.5.

tl;dr; I'm successfully working with UWP!
